src/v/cloud_storage/recursive_directory_walker.cc

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cloud_storage/recursive_directory_walker.h"
+
+#include "cloud_storage/logger.h"
+#include "utils/gate_guard.h"
+#include "vassert.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/file.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/core/sstring.hh>
+
+#include <algorithm>
+#include <chrono>
+
+namespace cloud_storage {
+
+ss::lowres_clock::time_point
+recursive_directory_walker::get_timepoint_from_stats(auto stats) {
+    auto last_access_sec = std::chrono::seconds(stats.st_atim.tv_sec);
+    auto last_access_nan = std::chrono::nanoseconds(stats.st_atim.tv_nsec);
+    auto last_access_timestamp = last_access_sec + last_access_nan;
+    auto last_access_timepoint = ss::lowres_clock::time_point(
+      std::chrono::duration_cast<ss::lowres_clock::duration>(
+        last_access_timestamp));
+
+    return last_access_timepoint;
+}
+
+ss::future<> recursive_directory_walker::walk(ss::sstring start_dir) {
+    gate_guard guard{_gate};
+
+    files.clear();
+    current_cache_size = 0;
+
+    std::deque<ss::sstring> dirlist = {start_dir};
+
+    while (!dirlist.empty()) {
+        auto target = dirlist.back();
+        dirlist.pop_back();
+        vassert(
+          0 == target.substr(0, start_dir.length()).compare(start_dir),
+          "Looking at directory {}, which is outside of initial dir {}.",
+          target,
+          start_dir);
+
+        try {
+            ss::file target_dir = co_await open_directory(target);
+            auto sub = target_dir.list_directory(
+              [this, &dirlist, target](
+                ss::directory_entry entry) -> ss::future<> {
+                  vlog(cst_log.debug, "Looking at directory {}", target);
+
+                  auto entry_path = std::filesystem::path(target)
+                                    / std::filesystem::path(entry.name);
+                  if (
+                    entry.type
+                    && entry.type == ss::directory_entry_type::regular) {
+                      vlog(cst_log.debug, "Regular file found {}", entry_path);
+
+                      ss::file f = co_await ss::open_file_dma(
```

Comment:
we can probably use a `seastar::reactor::file_stat` method and save an `open` syscall

Reply:
Used `ss::file_stats`